v1/users/factories/__init__.py

```diff
@@ -1 +1 @@
-from .user import UserFactory  # noqa: F401
\ No newline at end of file
+from .user import UserFactory  # noqa: F401
```

Comment:
having 
```
__all__ = ['UserFactory']
```
will allow you to drop # noqa: F401

Reply:
@vosi @buckyroberts yeah, thanks for review, ready to add this,
just want more of your opinion, what is the reason for adding ~4 and more lines of code in each folder,
does it makes any value? One of my principles is having less lines of code for many reasons, as well as less time taken for developers to read the code, and this lines (please correct me) is just allow to import this module with wildcard, like 'from .models import *' that i treat as anitpattern, and it not used in code anywhere?

Reply:
It's a kinda infrastructural, non-functional code, so I'd extend your principle, like "having fewer lines of FUNCTIONAL code." I believe the code should be additionally treated as documentation, so there are cases where it's better to be verbose.

You have a module (`factories`), and you don't want to expose that you have tons of files/submodules/classes/methods/etc there; you want to tell a user that here is the stuff that is ok to use, the rest - on one's own risk. So, you declare a kind of interface in the root __init__ listing stuff you want to expose.

In python, it's more a convention than language syntax.

Also, from module import * is not always an antipattern; it's ok to have it in __init__, listing children in __all__; there are other cases for this.

@fonar

Reply:
@vosi thank you for the explanation, just want to understand the usage of it.
For example if we look at 

/v1/teams/factories/__init__.py

we have

from .team import Team
from .team_member import TeamMember

doesn't that means that:
- we provide option to import this two classes directly from factories
- we inform devlopers that we are exposing exactly this two classes by this two lines of code

what additional value will give lines of code that contains __all__ ?
for me they looks like redundant here

Reply:
well, private and public keywords in Java are not redundant ...
same with \_\_all\_\_

Here's a good SO page abt it 
https://stackoverflow.com/questions/44834/can-someone-explain-all-in-python

Reply:
@vosi no, private and public words in Java are not redundant as private actually prevent class members to be accessed, and it makes it secure. In our case it's not.
In current our case this `__all__`  will not change anything or bring any value (as i understand it, please correct me if i'm wrong). Just trying to understand what vales will it bring and how it affect codebase and it's usage if we add this lines of code (until we planning to use wildcard imports _everywhere_, but i hope we not). if the case is to show developers what is exported,  it's already shown by two lines of code we already have in first two lines of code in `__init__.py`, and whats the reason to duplicate it

Reply:
On python.org in PEP-8 guide: https://www.python.org/dev/peps/pep-0008/#global-variable-names
`Modules that are designed for use via from M import * should use the __all__ mechanism to prevent exporting globals`
but we
1) don't have globals at this level for the cases: serializers, views, models, e.t.c.
2) not desgning for use import *
@vosi

Reply:
https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces

Reply:
dont use all if you dont like it, but in that case, remove redundant unused imports (the use noqa in this particular case cannot be justified)

Reply:
@vosi it's not about i like it or not, it's about redundant code that doesn't bring any value.
This imports are not redundant, and they was added by you in all folders, and they do their work:
they exporting models, serializers, factories, e.t.c. at the package level, as we have extended architecture, for example: instead of having one models.py file, we have models folder and one module per model.
So, this imports actually do the thing, but `__all__` variable do nothing.
But if you want i can remove this imports that will bring us back to @buckyroberts 's original architecture used in Bank and Validator.

Reply:
as of noqa, to avoid linter warnings there only two options add noqa F401 per line, or add it globally in flake8 config for __init__.py files

Reply:
Also, there's metion in PEP-8 styling guide "Wildcard imports (from <module> import *) should be avoided"

Reply:
there is another option to use \_\_all\_\_ which is a standard in Python community

Reply:
and to sum up, you are basically saying that private and public keywords in many programming languages are redundant
\_\_all\_\_ in python is a direct equivalent

Reply:
@vosi `__all__` in python have absolutely nothing in common with public/private/protected e.t.c. in C++, Java and other languages. public/private controls access to members of classes, and the only thing __all__ in python do (as we can see in documentation) is control what would be exported with wildcard imports (*) from module or package (and it is not recommended to use nowdays by pep-8)

according to standatard of comunity, i don't agree
It used in _some_ public libraries to expose list of interfaces that can be imported by wildcards.
and _never_ in every folder, every module, every file inside the project.

I just checked few of our dependency libraries and lot of them not usng `__all__` even for public library interfaces, examples:
rest_framework
django_channels
factory_boy
Faker

some of libraries just use "imports" in __init__ files to expose public interfaces of library at root packages level, and yeah, some provide __all__ for, but we know that it's only to allow wildcard imports.

My opinion that It's ok to add at least imports in `__init__.py` of root packages in thenewboston-python library, to show developers what classes is for external use, and also can be added `__all__` variable for develoeprs that will use thenewboston-python and will want to use wildcars imports.

But i still not heard any reason for adding it to project that is not a public library and to use it in every folder and every package or module of project thats only for internal use (as i said until we want to allow developers to use wildcard imports inside this projects)

Reply:
@vosi direct link to python documentation:
https://docs.python.org/3/tutorial/modules.html
```
if a package’s __init__.py code defines a list named __all__, it is taken to be the list of module names that should be imported when from package import * is encountered. It is up to the package author to keep this list up-to-date when a new version of the package is released. Package authors may also decide not to support it, if they don’t see a use for importing * from their package.
```

According to it, if we don't see a use for importing * from packages, we don't have to support `__all__` variable

Reply:
https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces

again, do the way you like, but redundant noqa is a sign of "code smell"
EITHER do not use imports in init, and use full path imports where you do an import, or use all

Reply:
Actually, we have 4 options
1) make it like in bank and validator projects, just leave `__init__.py` empty
2) use `__all__`
3) use #noqa
4) add rule for flake8 to ignore F402 for __init__.py
i'd better go with (3) or (4)

@buckyroberts what's your opinion?

Reply:
I do not consider 3 and 4 as options actually, its a noway, imo

Reply:
@fonar @vosi Let's go with 1 for now - make it like in Bank and Validator projects, just leave `__init__.py` empty

Reply:
@buckyroberts just discussed it with several django developers, they also think that (1) is better as it's easier to support project, and not require to update code in few places when changes was made.